### ISSUE TITLE
Segfaults in php-couchbase2 

### ISSUE THREAD

**Issue by a3mar** [remicollet/remirepo#136]
Package php-pecl-couchbase2-2.6.1-1.el7.remi.7.3.x86_64 was built with libcouchbase-2.10.3, but repo contains only 2.10.4 and 2.10.5 versions of libcouchbase:

```
~ # php -i |grep libcouch
libcouchbase runtime version => 2.10.5 (git: dabf6d58c2e89b4e75009a7f959e8ad30f85ff98)
libcouchbase headers version => 2.10.3 (git: b387722192f0bc49b0a846859610493e2f8129cd)

~ # yum list --show-duplicates libcouchbase
Available Packages
libcouchbase.x86_64   2.10.4-1.el7.remi    remi-safe 
libcouchbase.x86_64   2.10.5-1.el7.remi    remi-safe
```

After updating php-pecl-couchbase2 and libcouchbase to latest versions from repo, we get  segfaults in couchbase.so (from php-couchbase2) like this:

```
[Thu Jan 16 13:42:08 2020] php[18412]: segfault at 28 ip 00007f8255663e38 sp 00007fff5a4f6ff8 error 4 in couchbase.so[7f8255639000+5e000]
[Thu Jan 16 14:11:59 2020] php[19540]: segfault at 28 ip 00007fdeb2663e38 sp 00007fff360c44d8 error 4 in couchbase.so[7fdeb2639000+5e000]
```

After rebuilding php-pecl-couchbase2 segfaults fixed, please rebuild this package in repo with actual version of libcouchbase

**Comment by remicollet** [comment]
Strange, as ABI check seems OK
* [2.10.3 to 2.10.4](https://rpms.remirepo.net/compat_reports/libcouchbase/2.10.3_to_2.10.4/compat_report.html)
* [2.10.4 to 2.10.5](https://rpms.remirepo.net/compat_reports/libcouchbase/2.10.4_to_2.10.5/compat_report.html)

BTW, I will run a new set of build ASAP

@avsej Sergey, any idea of what may have happen ?

**Comment by remicollet** [comment]
Fixed by https://git.remirepo.net/cgit/rpms/php/pecl/php-pecl-couchbase2.git/commit/?id=93c5b8826b6079eb52b91b40d9a4c2cac3908d26

**Comment by remicollet** [comment]
Version 2.6.1-3 pushed to the repo.

**Comment by a3mar** [comment]
Thanks. 
But i reported this problem too soon, segfaults reproducing in new builds. I'll try to debug deeper.

**Comment by remicollet** [comment]
A gdb backtrace will be helpful

**Comment by a3mar** [comment]
Backtrace:
(gdb) bt
#0  0x00007f08289f2f38 in n1qlrow_callback (instance=0x55b2c05a89e0, ignoreme=<optimized out>, resp=0x7ffdf130a110) at /usr/src/debug/php-pecl-couchbase2-2.6.1/NTS/src/couchbase/bucket/n1ql.c:84
```
#1  0x00007f082873d7ef in invoke_row (is_last=true, resp=0x7ffdf130a110, this=0x55b2c05bf000) at /usr/src/debug/libcouchbase-2.10.5/src/n1ql/n1ql.cc:461
#2  ~lcb_N1QLREQ (this=0x55b2c05bf000, __in_chrg=<optimized out>) at /usr/src/debug/libcouchbase-2.10.5/src/n1ql/n1ql.cc:477
#3  lcb_N1QLREQ::~lcb_N1QLREQ (this=0x55b2c05bf000, __in_chrg=<optimized out>) at /usr/src/debug/libcouchbase-2.10.5/src/n1ql/n1ql.cc:501
#4  0x00007f08287024fc in timer_callback (sock=<optimized out>, which=<optimized out>, arg=0x55b2c05bf390) at /usr/src/debug/libcouchbase-2.10.5/src/lcbio/timer.c:45
#5  0x00007f0827a45a14 in event_process_active_single_queue (activeq=0x55b2c04a5020, base=0x55b2c05a8fc0) at event.c:1350
#6  event_process_active (base=<optimized out>) at event.c:1420
#7  event_base_loop (base=0x55b2c05a8fc0, flags=0) at event.c:1621
#8  0x00007f08287663f7 in lcb_wait (instance=0x55b2c05a89e0) at /usr/src/debug/libcouchbase-2.10.5/src/wait.cc:103
#9  0x00007f08289f33fa in pcbc_bucket_n1ql_request (bucket=bucket@entry=0x7f081d740980, cmd=0x7ffdf130a370, json_response=json_response@entry=1, json_options=json_options@entry=1, is_cbas=<optimized out>, return_value=0x7f083c21d6c0)
    at /usr/src/debug/php-pecl-couchbase2-2.6.1/NTS/src/couchbase/bucket/n1ql.c:178
#10 0x00007f08289ef1ee in zim_Bucket_query (execute_data=<optimized out>, return_value=0x7f083c21d6c0) at /usr/src/debug/php-pecl-couchbase2-2.6.1/NTS/src/couchbase/bucket.c:365
#11 0x000055b2bf1030ce in ZEND_DO_FCALL_SPEC_RETVAL_USED_HANDLER () at /usr/src/debug/php-7.3.13/Zend/zend_vm_execute.h:1106
#12 execute_ex (ex=0x4) at /usr/src/debug/php-7.3.13/Zend/zend_vm_execute.h:55521
#13 0x000055b2bf103873 in zend_execute (op_array=op_array@entry=0x7f083c2611c0, return_value=0x0, return_value@entry=0x7f0820f5d020) at /usr/src/debug/php-7.3.13/Zend/zend_vm_execute.h:60913
#14 0x000055b2bf074142 in zend_execute_scripts (type=type@entry=8, retval=0x7f0820f5d020, retval@entry=0x0, file_count=1008850512, file_count@entry=3) at /usr/src/debug/php-7.3.13/Zend/zend.c:1568
#15 0x000055b2bf013e50 in php_execute_script (primary_file=primary_file@entry=0x7ffdf130ca80) at /usr/src/debug/php-7.3.13/main/main.c:2639
#16 0x000055b2bee77a8b in main (argc=<optimized out>, argv=<optimized out>) at /usr/src/debug/php-7.3.13/sapi/fpm/fpm/fpm_main.c:1950
```

Looks like timeout from couchbase, with 0x0 as a pointer to htresp struct:
```
(gdb) info args
instance = 0x55b2c05a89e0
ignoreme = <optimized out>
resp = 0x7ffdf130a110

(gdb) p *((const lcb_RESPN1QL *) 0x7ffdf130a110)
$11 = {cookie = 0x7f081d5cee38, key = 0x0, nkey = 0, cas = 0, rc = LCB_ETIMEDOUT, version = 0, rflags = 1, row = 0x7f082d60e3f8 <std::string::_Rep::_S_empty_rep_storage+24> "", nrow = 0, htresp = 0x0}

(gdb) p *((const lcb_RESPN1QL *) 0x7ffdf130a110)->htresp
Cannot access memory at address 0x0
```

In https://github.com/couchbase/php-couchbase/blob/release-2.6/src/couchbase/bucket/n1ql.c#L84 module trying to access htstatus field of this struct and we got segfault. I'll create new issue in module repo.

**Comment by remicollet** [comment]
So, looks like a upstream bug, can you please file a bug there ? /cc  @avsej

**Comment by a3mar** [comment]
I dont know how to report bugs in couchbase, their github without issues feature. We're planning to buy commercial support from them and i sent link to this issue to support. Then, i close it here.

**Comment by remicollet** [comment]
@a3mar see https://github.com/couchbase/php-couchbase/pull/23

waiting for upstream feedback

BTW, libcouchbase version 3 is released, and pecl ext new version should be released soon (beta1 was released in Oct), so I don't know if they will release a 2.6.2, or will prefer to concentrate on v3.

**Comment by remicollet** [comment]
FYI, as upstream have merged my fix proposal, I run a new set of builds (2.6.1-4) which includes the fix.

**Comment by a3mar** [comment]
Thank you

**Comment by avsej** [comment]
I've released 2.6.2 version today, which includes this fix. Thank you.
